File: R/nat.fastcore/R/warp.R
# Landmark-based spatial transforms: thin-plate spline and moving least squares.

# Validate a landmark pair R-side and return them as (M, 3) doubles.
#
# extendr 0.7 loses a panic's payload when it is raised from an *associated* function, so a
# constructor that failed in Rust would reach R as "User function panicked: fit". Everything
# we can check up front therefore gets checked here; the Rust side reports what it can only
# discover by trying (a singular system) through `error()`.
.warp_landmarks <- function(source, target, min_n = 1L, what = "landmark") {
  src <- .xform_xyz(source, "source")
  trg <- .xform_xyz(target, "target")

  if (nrow(src) != nrow(trg)) {
    stop(sprintf(
      "number of source landmarks (%d) must match number of target landmarks (%d)",
      nrow(src), nrow(trg)
    ))
  }
  if (nrow(src) < min_n) {
    stop(sprintf(
      "need at least %d %ss, got %d", min_n, what, nrow(src)
    ))
  }
  if (!all(is.finite(src)) || !all(is.finite(trg))) {
    stop("landmarks must not contain NA, NaN or Inf")
  }
  list(source = src, target = trg)
}

# Raise the Rust-side failure, if any, as a real R condition.
.warp_check <- function(ptr) {
  err <- ptr$error()
  if (nzchar(err)) stop(err, call. = FALSE)
  ptr
}

.tps_ptr <- function(reg) {
  if (!inherits(reg, "tps_transform")) {
    stop("`reg` must be a <tps_transform>, as returned by `tps_transform()`")
  }
  reg$ptr
}

.mls_ptr <- function(reg) {
  if (!inherits(reg, "mls_transform")) {
    stop("`reg` must be an <mls_transform>, as returned by `mls_transform()`")
  }
  reg$ptr
}

#' Fit a thin-plate spline to landmark pairs
#'
#' A thin-plate spline interpolates the source landmarks onto the target landmarks exactly,
#' and between them minimises the integral bending norm -- the smoothest warp consistent
#' with the landmarks. This is the usual fallback when no image registration exists for a
#' pair of template spaces, and what `nat`'s `tpsreg` provides.
#'
#' The fit happens once, here, and the resulting object can then be applied as often as you
#' like with [tps_xform()]. That split matters: the fit is *cubic* in the landmark count
#' while applying it is linear, so refitting per call would dominate any real workload.
#'
#' There is no `batch_size` to tune, unlike the `navis` implementation this mirrors. The
#' `(n_points, n_landmarks)` distance matrix is fused into the accumulation rather than
#' built, so peak memory is the output matrix regardless of how many points or landmarks are
#' involved.
#'
#' @param source An `(M, 3)` matrix of source landmarks. A data frame is also accepted -- if
#'   it has x/y/z columns (in any case) those are used, whatever else it carries, otherwise
#'   the first three columns.
#' @param target An `(M, 3)` matrix of target landmarks, one per source landmark.
#'
#' @return An object of class `tps_transform`.
#'
#' @section Inverting:
#' A thin-plate spline has no closed-form inverse. To map the other way, fit the other way
#' round -- `tps_transform(target, source)` -- which is a fresh fit, not an inversion.
#'
#' @examples
#' src <- rbind(c(0, 0, 0), c(10, 10, 10), c(100, 100, 100), c(80, 10, 30))
#' trg <- rbind(c(1, 15, 5), c(9, 18, 21), c(80, 99, 120), c(5, 10, 80))
#' tps <- tps_transform(src, trg)
#' tps
#'
#' # landmarks are reproduced exactly
#' stopifnot(all.equal(tps_xform(tps, src), trg))
#'
#' tps_xform(tps, rbind(c(0, 0, 0), c(50, 50, 50)))
#'
#' @seealso [tps_xform()], [mls_transform()]
#' @export
tps_transform <- function(source, target) {
  lm <- .warp_landmarks(source, target, min_n = 4L, what = "landmark")
  ptr <- .warp_check(TpsTransformPtr$fit(lm$source, lm$target))
  structure(
    list(ptr = ptr, source = lm$source, target = lm$target),
    class = "tps_transform"
  )
}

#' @export
print.tps_transform <- function(x, ...) {
  cat(sprintf("<tps_transform> %d landmarks\n", x$ptr$n_landmarks()))
  invisible(x)
}

#' Apply a thin-plate spline to points
#'
#' @param reg A `tps_transform`, from [tps_transform()].
#' @param xyz An `(N, 3)` matrix of coordinates. A data frame is also accepted -- if it has
#'   x/y/z columns (in any case) those are used, whatever else it carries, otherwise the
#'   first three columns -- as is a bare length-3 vector. `nat::xyzmatrix()` gives you this
#'   for a neuron.
#' @param n_cores Number of threads. `NULL` (default) uses all cores.
#'
#' @return An `(N, 3)` matrix of transformed coordinates.
#'
#' @examples
#' src <- rbind(c(0, 0, 0), c(10, 10, 10), c(100, 100, 100), c(80, 10, 30))
#' trg <- rbind(c(1, 15, 5), c(9, 18, 21), c(80, 99, 120), c(5, 10, 80))
#' tps <- tps_transform(src, trg)
#' tps_xform(tps, c(50, 50, 50))
#'
#' \dontrun{
#' # transform a neuron
#' n <- nat::Cell07PNs[[1]]
#' nat::xyzmatrix(n) <- tps_xform(tps, nat::xyzmatrix(n))
#' }
#'
#' @seealso [tps_transform()], [tps_affine()]
#' @export
tps_xform <- function(reg, xyz, n_cores = NULL) {
  .tps_ptr(reg)$xform(
    .xform_xyz(xyz),
    if (is.null(n_cores)) NULL else as.integer(n_cores)
  )
}

#' The affine part of a thin-plate spline
#'
#' A thin-plate spline decomposes into an affine part and a non-affine ("bending") part.
#' This returns the affine, which is what the spline converges to far from the landmarks.
#'
#' @param reg A `tps_transform`, from [tps_transform()].
#'
#' @return A `(4, 4)` homogeneous transformation matrix, with last row `c(0, 0, 0, 1)`.
#'
#' @examples
#' # landmarks related by a pure affine give back exactly that affine
#' src <- rbind(c(0, 0, 0), c(1, 0, 0), c(0, 1, 0), c(0, 0, 1), c(1, 1, 1))
#' trg <- src * 2 + matrix(c(3, 4, 5), nrow(src), 3, byrow = TRUE)
#' tps_affine(tps_transform(src, trg))
#'
#' @seealso [tps_transform()], [mls_affine()]
#' @export
tps_affine <- function(reg) {
  .tps_ptr(reg)$matrix_affine()
}

#' Coefficients of a fitted thin-plate spline
#'
#' Returns the fit itself: `W`, the weights of the non-affine part, and `A`, the
#' coefficients of the affine part. Together with the source landmarks these fully determine
#' the transform, so they are what you serialise if you want to store a fit rather than
#' repeat it.
#'
#' @param reg A `tps_transform`, from [tps_transform()].
#'
#' @return A list with elements `source` (`(M, 3)`), `W` (`(M, 3)`) and `A` (`(4, 3)`, whose
#'   first row is the translation).
#'
#' @examples
#' src <- rbind(c(0, 0, 0), c(10, 10, 10), c(100, 100, 100), c(80, 10, 30))
#' trg <- rbind(c(1, 15, 5), c(9, 18, 21), c(80, 99, 120), c(5, 10, 80))
#' co <- tps_coefs(tps_transform(src, trg))
#' dim(co$W)
#' dim(co$A)
#'
#' @seealso [tps_transform()]
#' @export
tps_coefs <- function(reg) {
  ptr <- .tps_ptr(reg)
  list(source = ptr$source(), W = ptr$weights(), A = ptr$affine_coefs())
}

#' Build a moving least squares transform from landmark pairs
#'
#' The affine flavour of the algorithm published in Schaefer et al. (2006). Unlike a
#' thin-plate spline there is no fit: every point gets its *own* affine, solved on the fly
#' from all landmarks weighted by inverse squared distance. Construction is therefore free
#' and [mls_xform()] is the entire cost.
#'
#' Results are similar to but not identical with [tps_transform()]; which suits a given
#' pair of spaces is worth checking empirically.
#'
#' @param source An `(M, 3)` matrix of source landmarks. A data frame is also accepted -- if
#'   it has x/y/z columns (in any case) those are used, whatever else it carries, otherwise
#'   the first three columns.
#' @param target An `(M, 3)` matrix of target landmarks, one per source landmark.
#' @param direction `"forward"` (default), or `"inverse"` to treat the target as the source
#'   and vice versa. Note this fits the warp in the opposite direction; it is not an exact
#'   inverse, which moving least squares does not have. It is free, though -- no refit.
#'
#' @return An object of class `mls_transform`.
#'
#' @references
#' Schaefer S, McPhail T, Warren J (2006). "Image deformation using moving least squares."
#' *ACM Transactions on Graphics* 25(3):533-540. \doi{10.1145/1141911.1141920}
#'
#' @examples
#' src <- rbind(c(0, 0, 0), c(10, 10, 10), c(100, 100, 100), c(80, 10, 30))
#' trg <- rbind(c(1, 15, 5), c(9, 18, 21), c(80, 99, 120), c(5, 10, 80))
#' mls <- mls_transform(src, trg)
#' mls
#'
#' # a point sitting on a landmark maps to its partner
#' stopifnot(all.equal(mls_xform(mls, src), trg, tolerance = 1e-6))
#'
#' # and the inverse direction maps them back
#' back <- mls_transform(src, trg, direction = "inverse")
#' stopifnot(all.equal(mls_xform(back, trg), src, tolerance = 1e-6))
#'
#' @seealso [mls_xform()], [tps_transform()]
#' @export
mls_transform <- function(source, target, direction = c("forward", "inverse")) {
  direction <- match.arg(direction)
  lm <- .warp_landmarks(source, target, min_n = 1L, what = "landmark")
  ptr <- .warp_check(MlsTransformPtr$build(lm$source, lm$target))
  structure(
    list(
      ptr = ptr, source = lm$source, target = lm$target,
      direction = direction, reverse = direction == "inverse"
    ),
    class = "mls_transform"
  )
}

#' @export
print.mls_transform <- function(x, ...) {
  cat(sprintf(
    "<mls_transform> %d landmarks, direction '%s'\n",
    x$ptr$n_landmarks(), x$direction
  ))
  invisible(x)
}

#' Apply a moving least squares transform to points
#'
#' @param reg An `mls_transform`, from [mls_transform()].
#' @param xyz An `(N, 3)` matrix of coordinates. A data frame is also accepted -- if it has
#'   x/y/z columns (in any case) those are used, whatever else it carries, otherwise the
#'   first three columns -- as is a bare length-3 vector. `nat::xyzmatrix()` gives you this
#'   for a neuron.
#' @param n_cores Number of threads. `NULL` (default) uses all cores.
#'
#' @return An `(N, 3)` matrix of transformed coordinates.
#'
#' @examples
#' src <- rbind(c(0, 0, 0), c(10, 10, 10), c(100, 100, 100), c(80, 10, 30))
#' trg <- rbind(c(1, 15, 5), c(9, 18, 21), c(80, 99, 120), c(5, 10, 80))
#' mls <- mls_transform(src, trg)
#' mls_xform(mls, c(50, 50, 50))
#'
#' @seealso [mls_transform()], [mls_affine()]
#' @export
mls_xform <- function(reg, xyz, n_cores = NULL) {
  .mls_ptr(reg)$xform(
    .xform_xyz(xyz),
    isTRUE(reg$reverse),
    if (is.null(n_cores)) NULL else as.integer(n_cores)
  )
}

#' The global affine of a moving least squares transform
#'
#' Moving least squares is *locally* weighted -- every point effectively gets its own affine
#' -- so no single matrix describes it. This returns the *global* affine: the least-squares
#' fit of the source landmarks onto the target landmarks, which is what the warp converges
#' to far from them, where the distance weights even out.
#'
#' @param reg An `mls_transform`, from [mls_transform()].
#'
#' @return A `(4, 4)` homogeneous transformation matrix, with last row `c(0, 0, 0, 1)`.
#'
#' @examples
#' src <- rbind(c(0, 0, 0), c(1, 0, 0), c(0, 1, 0), c(0, 0, 1), c(1, 1, 1))
#' trg <- src * 2 + matrix(c(3, 4, 5), nrow(src), 3, byrow = TRUE)
#' mls_affine(mls_transform(src, trg))
#'
#' @seealso [mls_transform()], [tps_affine()]
#' @export
mls_affine <- function(reg) {
  .mls_ptr(reg)$matrix_affine(isTRUE(reg$reverse))
}

File: py/navis_fastcore/warp.py
"""Landmark-based spatial transforms: thin-plate spline and moving least squares.

These are the fallback when no image registration exists for a pair of template spaces -
you have matched landmarks and want a warp that interpolates between them.

Both are drop-in replacements for their `navis` counterparts
(``navis.transforms.TPStransform`` and ``navis.transforms.MovingLeastSquaresTransform``)
and agree with them to ~1e-14 relative. The difference is that the ``(n_points,
n_landmarks)`` matrix the reference implementations build is never materialised here, so
there is no ``batch_size`` to tune and peak memory does not depend on the landmark count.
"""

import numpy as np

from . import _fastcore
from ._points import _prep_points

__all__ = ["TpsTransform", "MlsTransform"]


def _prep_xyz(points, name="points"):
    """Coerce to a C-contiguous (N, 3) float64 array, accepting a DataFrame.

    `navis`' landmark transforms take x/y/z-columned DataFrames, and landmarks are
    routinely read straight out of a CSV, so accepting one here is what makes these classes
    drop-in. Duck-typed on `.columns` rather than importing pandas, which is not a
    dependency of this package.
    """
    cols = getattr(points, "columns", None)
    if cols is not None:
        missing = [c for c in ("x", "y", "z") if c not in cols]
        if missing:
            raise ValueError(
                f"`{name}` DataFrame must have x/y/z columns, missing: "
                f"{', '.join(missing)}"
            )
        points = points[["x", "y", "z"]].values
    return _prep_points(points, name=name)


def _prep_landmarks(source, target):
    """Coerce a landmark pair to matching C-contiguous (M, 3) float64 arrays."""
    src, _ = _prep_xyz(source, name="landmarks_source")
    trg, _ = _prep_xyz(target, name="landmarks_target")
    if src.shape[0] != trg.shape[0]:
        raise ValueError(
            "number of source landmarks must match number of target landmarks: got "
            f"{src.shape[0]} and {trg.shape[0]}"
        )
    return src, trg


class TpsTransform:
    """A thin-plate spline transform, fitted to landmark pairs.

    The spline interpolates the source landmarks onto the target landmarks exactly, and
    between them it minimises the integral bending norm - the smoothest warp consistent
    with the landmarks.

    The fit happens once, here, and the transform can then be applied any number of times.

    Parameters
    ----------
    landmarks_source :  (M, 3) array-like
                        Source landmarks as x/y/z coordinates. A pandas ``DataFrame`` with
                        x/y/z columns is also accepted.
    landmarks_target :  (M, 3) array-like
                        Target landmarks, one per source landmark.

    Attributes
    ----------
    source :            (M, 3) np.ndarray
                        The source landmarks.
    W :                 (M, 3) np.ndarray
                        Weights of the non-affine part of the spline.
    A :                 (4, 3) np.ndarray
                        Coefficients of the affine part; row 0 is the translation.
    matrix_affine :     (4, 4) np.ndarray
                        The affine part as a homogeneous matrix - what the spline
                        converges to far from the landmarks.

    Notes
    -----
    Unlike ``navis.transforms.TPStransform`` there is no ``batch_size``: the distance
    matrix is fused into the accumulation rather than built, so peak memory is the output
    array regardless of how many points or landmarks are involved.

    The fit is cubic in the landmark count and runs through a blocked LU. At a few thousand
    landmarks it is somewhat slower than ``numpy.linalg.solve`` (which reaches hardware
    LAPACK); :meth:`from_coefs` exists if you would rather fit with numpy and only use this
    class to apply the result.

    Examples
    --------
    >>> import navis_fastcore as fastcore
    >>> import numpy as np
    >>> src = np.array([[0, 0, 0], [10, 10, 10], [100, 100, 100], [80, 10, 30]])
    >>> trg = np.array([[1, 15, 5], [9, 18, 21], [80, 99, 120], [5, 10, 80]])
    >>> tr = fastcore.TpsTransform(src, trg)
    >>> tr.xform(np.array([[0, 0, 0], [50, 50, 50]]))
    array([[ 1.        , 15.        ,  5.        ],
           [40.55555556, 54.        , 65.        ]])
    >>> # Landmarks are reproduced exactly
    >>> np.allclose(tr.xform(src), trg)
    True
    >>> # Negation refits in the opposite direction
    >>> np.allclose((-tr).xform(trg), src)
    True

    """

    def __init__(self, landmarks_source, landmarks_target):
        src, trg = _prep_landmarks(landmarks_source, landmarks_target)
        if src.shape[0] < 4:
            raise ValueError(
                "need at least 4 landmarks to fit a 3D thin-plate spline, got "
                f"{src.shape[0]}"
            )
        self._target = trg
        self._tr = _fastcore.TpsTransform(src, trg)

    @classmethod
    def from_coefs(cls, landmarks_source, W, A, landmarks_target=None):
        """Build from coefficients fitted elsewhere, skipping the fit.

        Parameters
        ----------
        landmarks_source :  (M, 3) array-like
                            The source landmarks the coefficients belong to.
        W :                 (M, 3) array-like
                            Weights of the non-affine part.
        A :                 (4, 3) array-like
                            Coefficients of the affine part.
        landmarks_target :  (M, 3) array-like, optional
                            The target landmarks. Not needed to transform points, but
                            without them :meth:`__neg__` cannot refit the inverse.

        Returns
        -------
        TpsTransform

        Examples
        --------
        >>> import navis_fastcore as fastcore
        >>> import numpy as np
        >>> src = np.array([[0, 0, 0], [10, 10, 10], [100, 100, 100], [80, 10, 30]])
        >>> trg = np.array([[1, 15, 5], [9, 18, 21], [80, 99, 120], [5, 10, 80]])
        >>> tr = fastcore.TpsTransform(src, trg)
        >>> same = fastcore.TpsTransform.from_coefs(tr.source, tr.W, tr.A)
        >>> np.allclose(same.xform(src), trg)
        True

        """
        obj = cls.__new__(cls)
        src, _ = _prep_xyz(landmarks_source, name="landmarks_source")
        w = np.ascontiguousarray(np.asarray(W, dtype=np.float64))
        a = np.ascontiguousarray(np.asarray(A, dtype=np.float64))
        if w.shape != src.shape:
            raise ValueError(f"`W` must be (M, 3) matching the landmarks, got {w.shape}")
        if a.shape != (4, 3):
            raise ValueError(f"`A` must be (4, 3), got {a.shape}")
        obj._tr = _fastcore.TpsTransform.from_coefs(src, w, a)
        obj._target = (
            None
            if landmarks_target is None
            else _prep_xyz(landmarks_target, name="landmarks_target")[0]
        )
        return obj

    def xform(self, points, n_cores=None):
        """Transform points.

        Parameters
        ----------
        points :    (N, 3) array-like
                    Coordinates to transform. A single ``(3,)`` point is accepted and
                    returns a ``(3,)`` point; a ``DataFrame`` with x/y/z columns also works.
        n_cores :   int, optional
                    Number of threads. ``None`` (default) uses all cores.

        Returns
        -------
        (N, 3) np.ndarray
                    Transformed coordinates.

        """
        pts, was_1d = _prep_xyz(points)
        out = self._tr.xform(pts, None if n_cores is None else int(n_cores))
        return out[0] if was_1d else out

    @property
    def source(self):
        """The source landmarks, as an (M, 3) array."""
        return self._tr.source

    @property
    def target(self):
        """The target landmarks, or ``None`` if built from coefficients without them."""
        return None if self._target is None else self._target.copy()

    @property
    def W(self):
        """Weights of the non-affine part of the spline, as an (M, 3) array."""
        return self._tr.weights

    @property
    def A(self):
        """Coefficients of the affine part, as a (4, 3) array."""
        return self._tr.affine_coefs

    @property
    def matrix_affine(self):
        """The affine part as a (4, 4) homogeneous matrix."""
        return self._tr.matrix_affine

    def copy(self):
        """Return a copy. Shares the fit rather than repeating it."""
        obj = self.__class__.__new__(self.__class__)
        obj._tr = self._tr
        obj._target = None if self._target is None else self._target.copy()
        return obj

    def __neg__(self):
        """Fit the spline in the opposite direction.

        This is a fresh fit of target onto source, not an inversion of this one - a thin
        plate spline has no closed-form inverse. Requires the target landmarks.
        """
        if self._target is None:
            raise ValueError(
                "cannot invert a TpsTransform built via `from_coefs` without "
                "`landmarks_target`"
            )
        return self.__class__(self._target, self.source)

    def __eq__(self, other):
        if not isinstance(other, TpsTransform):
            return NotImplemented
        if not np.array_equal(self.source, other.source):
            return False
        if self._target is None or other._target is None:
            # No targets to compare - fall back to the coefficients, which determine the
            # transform just as completely.
            return np.array_equal(self.W, other.W) and np.array_equal(self.A, other.A)
        return np.array_equal(self._target, other._target)

    def __len__(self):
        return self._tr.n_landmarks

    def __reduce__(self):
        # Ship the coefficients, not the landmarks alone: refitting in every
        # multiprocessing worker would cost more than the transforms they were forked to
        # perform.
        return (
            _tps_from_state,
            (self.source, self.W, self.A, self._target),
        )

    def __repr__(self):
        return f"<TpsTransform(landmarks={len(self)})>"


def _tps_from_state(source, W, A, target):
    """Unpickle helper - module-level so it is importable by name."""
    return TpsTransform.from_coefs(source, W, A, landmarks_target=target)


class MlsTransform:
    """A moving least squares transform, defined by landmark pairs.

    The affine flavour of the algorithm published in
    [Schaefer et al. (2006)](https://dl.acm.org/doi/pdf/10.1145/1179352.1141920). Unlike a
    thin-plate spline there is no fit: every point gets its *own* affine, solved on the fly
    from all landmarks weighted by inverse squared distance. That makes construction free
    and :meth:`xform` the entire cost.

    Parameters
    ----------
    landmarks_source :  (M, 3) array-like
                        Source landmarks as x/y/z coordinates. A pandas ``DataFrame`` with
                        x/y/z columns is also accepted.
    landmarks_target :  (M, 3) array-like
                        Target landmarks, one per source landmark.
    direction :         "forward" | "inverse"
                        ``"inverse"`` treats the target as the source and vice versa.
                        Note this fits the warp in the opposite direction; it is not an
                        exact inverse, which moving least squares does not have.

    Notes
    -----
    ``navis.transforms.MovingLeastSquaresTransform`` (via `molesq`) builds
    ``(3, M, N)``-shaped intermediates, which is why it takes a ``batch_size`` - and why in
    practice it cannot run at the landmark counts real registrations use: 3400 landmarks at
    the default batch size needs ~23 GB. Here everything but the result is a reduction over
    landmarks, so peak memory is the output array and the landmark count is unbounded.

    Examples
    --------
    >>> import navis_fastcore as fastcore
    >>> import numpy as np
    >>> src = np.array([[0, 0, 0], [10, 10, 10], [100, 100, 100], [80, 10, 30]])
    >>> trg = np.array([[1, 15, 5], [9, 18, 21], [80, 99, 120], [5, 10, 80]])
    >>> tr = fastcore.MlsTransform(src, trg)
    >>> # A point sitting on a landmark maps to its partner
    >>> np.allclose(tr.xform(src), trg)
    True
    >>> # Negation swaps the direction
    >>> np.allclose((-tr).xform(trg), src)
    True

    """

    def __init__(self, landmarks_source, landmarks_target, direction="forward"):
        if direction not in ("forward", "inverse"):
            raise ValueError(
                f'`direction` must be "forward" or "inverse", got {direction!r}'
            )
        src, trg = _prep_landmarks(landmarks_source, landmarks_target)
        self._reverse = direction == "inverse"
        self._tr = _fastcore.MlsTransform(src, trg)

    def xform(self, points, n_cores=None, reverse=None):
        """Transform points.

        Parameters
        ----------
        points :    (N, 3) array-like
                    Coordinates to transform. A single ``(3,)`` point is accepted and
                    returns a ``(3,)`` point; a ``DataFrame`` with x/y/z columns also works.
        n_cores :   int, optional
                    Number of threads. ``None`` (default) uses all cores.
        reverse :   bool, optional
                    Override this transform's ``direction`` for this call only:
                    ``False`` maps source -> target, ``True`` target -> source.
                    ``None`` (default) uses ``direction``. The underlying fit is
                    direction-agnostic, so this saves rebuilding the object just
                    to run it the other way.

        Returns
        -------
        (N, 3) np.ndarray
                    Transformed coordinates.

        """
        pts, was_1d = _prep_xyz(points)
        rev = self._reverse if reverse is None else bool(reverse)
        out = self._tr.xform(pts, rev, None if n_cores is None else int(n_cores))
        return out[0] if was_1d else out

    @property
    def source(self):
        """The landmarks this transform maps *from*, honouring ``direction``."""
        return self._tr.target if self._reverse else self._tr.source

    @property
    def target(self):
        """The landmarks this transform maps *to*, honouring ``direction``."""
        return self._tr.source if self._reverse else self._tr.target

    @property
    def direction(self):
        """``"forward"`` or ``"inverse"``."""
        return "inverse" if self._reverse else "forward"

    @property
    def matrix_affine(self):
        """The *global* affine as a (4, 4) homogeneous matrix.

        Moving least squares is *locally* weighted - every point effectively gets its own
        affine - so there is no single matrix describing it. This is the least-squares fit
        of source onto target landmarks, which is what the warp converges to far from them.
        """
        return self._tr.matrix_affine(self._reverse)

    def copy(self):
        """Return a copy."""
        obj = self.__class__.__new__(self.__class__)
        obj._tr = self._tr
        obj._reverse = self._reverse
        return obj

    def __neg__(self):
        """Flip the direction."""
        obj = self.copy()
        obj._reverse = not self._reverse
        return obj

    def __eq__(self, other):
        if not isinstance(other, MlsTransform):
            return NotImplemented
        return (
            np.array_equal(self.source, other.source)
            and np.array_equal(self.target, other.target)
        )

    def __len__(self):
        return self._tr.n_landmarks

    def __reduce__(self):
        return (
            _mls_from_state,
            (self._tr.source, self._tr.target, self.direction),
        )

    def __repr__(self):
        return f"<MlsTransform(landmarks={len(self)}, direction='{self.direction}')>"


def _mls_from_state(source, target, direction):
    """Unpickle helper - module-level so it is importable by name."""
    return MlsTransform(source, target, direction=direction)
